### ISSUE TITLE
Basic Reports

### ISSUE THREAD

**Issue by jterps08** [PowerlineApp/powerline-mobile#410]
From #408 

This actually raises the broader issue of reporting results for all items:

**(In Scope)**

1. Membership Roster
2. Poll Responses
3. Petition Responses
4. Fundraiser Responses
5. Event Responses

Responses include names (if response not private), response/answer choice, comments, and required information fields.
Membership roster includes user profile information (only permissioned-information) plus required information fields.

**(Out of Scope)**

1. User Post Responses - Ability for people (not just leaders) to review the responses to a user's post, anonymized, by general user profile information and linked electeds.
2. Analytics on Items 2-5 above - Ability for people to cross-tabulate information from user profile and connected leaders/districts with responses to polls/petitions/fundraisers/event/posts.

So... The question is how can we achieve the in-scope objectives above as easily as possible? XLS downloads for now?

**Comment by jterps08** [comment]
continuing this conversation: https://github.com/PowerlineApp/powerline-mobile/issues/408#issuecomment-258356950

Thanks @petervojtek - So, let's try to get the "Export Data" button in the app once Igor has it ready. Please proceed on other items in the mean time... 

 @igaponov - It looks like I/we forgot a major item... We need to be able to provide reports as per the In-Scope list above. Let's start with the Membership roster. Currently this is in production as an XLS export. We need the same available in dev. Peter will need to create an "Export Roster" button in the app that calls on the API to return the roster for a specific group, assuming he has the authority to do it. 

Once we complete roster, we'll do a similar exercise for polls, leader petitions, fundraisers, and events. 

The thing to keep in mind here: 1) group permissions -- what is the leader allowed to see based on what permissions the user asked for, and 2) anonymous responses - when a user submits an anonymous response to a poll or fundraiser, then we can't show his name in the report...

**Comment by igaponov** [comment]
@jterps08 user profile information (only permissioned-information) is clear, but what is "required information fields"? Can you please show an example of one row from xls file?

**Comment by jterps08** [comment]
please see google drive xls

**Comment by jterps08** [comment]
_This is effectively the initial iteration of our Data Visualization & Analytics capability_ - there will be other iterations for certain down the road...

@igaponov Here's how I think we should do this... Basically, I think we need to have a data reporting endpoint that we can call for XLS reports with certain parameters to determine the type of report and for what type of information... Eventually, we will move beyond this XLS files concept and we will return permission-ed data back for D3.js -driven visual display and manipulation (out-of-scope for now)... Please keep this in mind from a design consideration perspective. We will plan to continue to provide XLS file reports in the future in addition to providing API access to permissioned data for visualization and analytics purposes. 

@austinpapp - I think we can safely execute on the in-scope element of this effort (XLS file reports) using our existing server, but if you think we should build this from the beginning as a standalone application on a different server / python / etc as we've tangentially discussed previously, please comment now. From a load balancing standpoint in the future, I think all data endpoint requests can go to a dedicated server. I think we are safe for doing XLS file reports on here for now since they're relatively simple to produce, but I think we'll want our own standalone reporting application to access datawarehouse / cloned/shadowed/whatever databases in the future...

@petervojtek 
From a frontend perspective, we'll need an icon (something similar to this: http://www.redtone.com/wp-content/uploads/2015/05/reporticon-blue1.png) that will be visible to leaders only that allows them to download a report for the relevant item(s) in their group(s)... I think this icon will have to go on the item detail screen because I'm not sure we have enough room in the newsfeed quick action bar, which is fine.

When the leader taps on this icon for an item in his group (poll/petition/event/fundraiser), a pop-up displays "Download report?" (Download, Cancel)... 

Similarly, in the "Manager Group" screen, I think we need to add the button "Download Group Roster" beneath the name of the group on the top of the screen (not inside a drawer), similarly with a popup "Download membership roster?" (Download, Cancel).

**Comment by igaponov** [comment]
@jterps08 what does `group karma amount` mean?

**Comment by jterps08** [comment]
You can ignore it or leave it blank for now... When we implement points/karma, this is where we will display how much karma the user has earned from activities s/he has done inside the group itself.

**Comment by igaponov** [comment]
`GET /api/v2/groups/{id}/members`
Use header `Accept: text/csv` to get csv file. Without this header the endpoint will return json.

**Comment by jterps08** [comment]
Cool - so thats for the membership roster report... Will the approach above work for getting the report for a poll or petition or fundraiser or event?

**Comment by igaponov** [comment]
@jterps08 what does `required information fields` mean for response reports? I see only answer and comment in your table.

**Comment by jterps08** [comment]
Sorry for confusion @igaponov -- http://52.5.45.80:8080/#/v2-group-fields is what I'm referring to (there may be 0, 1, or many fields)

**Comment by igaponov** [comment]
@jterps08 I think you confused with Membership roster, there is no sense to add group fields to poll responses.

**Comment by jterps08** [comment]
@igaponov Columns B-Q should be included in every report. Only Items B-Q are in membership roster... 

For a single and specific poll/petition/event/fundraiser report, for example, the report is Columns B through Q + Column S + Column T.... e.g. for Poll 255. 

For group's poll report, the report is Columns B through Q + Columns V through AE... e.g. For Poll 255, Poll 256, Poll 257 (if the group only had 3 polls for example)... 

Does that make any sense? Sorry if I'm not clear - I may not have though thtis through correctly... Columns V-Z are really valuable, but harder...

**Comment by jterps08** [comment]
For what it's worth, the value of the data by combining all of these data points into one data set allows to gain some extraordinary insights... e.g. 55% of the people who answered the poll with Answer A also have Humpbacks as their favorite whale... 

Columns V-Z are mostly important for Posts and Petitions (not polls/fundraisers/events), but it's a good exercise nonetheless for this initial round...

**Comment by igaponov** [comment]
@jterps08 what is `single and specific poll report` and what is the difference between it and `group's poll report`? I don't see it in your first comment. 
Columns V-Z belong to representatives, right?

**Comment by jterps08** [comment]
Columns V-Z belongs to linked representatives yes.

For single/specific poll report, this is the ability of the endpoint to produce a report that includes Clumns B through Q + Column S + Column T where S is the answer choice only for the specific/single poll requested (e.g. Poll 25.... This is the higher priority item.

For group poll report, this is B through Q + Columns V through AE which includes ALL of the group's polls and their answer choices as individual columns... So if there's 50 members in the group, there will be 50 rows (not including header). If there's 15 polls that the group has published, there will be 15 columns after B through Q...

**Comment by igaponov** [comment]
~~@jterps08 will info in columns V-Z be the same for all 50 users?~~

Ok, I see. It'll be different if users are in different districts.

**Comment by igaponov** [comment]
`GET /api/v2/polls/{id}/responses`

I'll add group poll report after representative refactoring

**Comment by jterps08** [comment]
@igaponov I know you finished https://github.com/PowerlineApp/powerline-server/issues/244... Which needed to be done per your comment above before proceeding... please update Peter once you have completed the reports so that he can incorporate on frontend

**Comment by igaponov** [comment]
@jterps08 PowerlineApp/powerline-server#244 is not finished. Group's poll report can be done only after 244. Other reports are ready.

**Comment by igaponov** [comment]
@jterps08 you said: 
> anonymous responses - when a user submits an anonymous response to a poll or fundraiser, then we can't show his name in the report...

But there are several responses in group poll report and they can have different privacy settings.

Also, I'm not sure how to attach representative to this report. Columns are constant, I should know them BEFORE running a query. For example, I know all group's fields and polls so can join them. It's looks like you want to get positions/posts (senator, congressman, mayor etc), but we don't have this kind of link between a group and representatives.

**Comment by jterps08** [comment]
@igaponov If the user responded with privacy setting on, you can replace their name with "(Anonymous)" in the report. If the user doesn't have the privacy setting on, you provide the actual information. It may also be necessary to use this on the other fields (like phone address and e-mail) in order to protect the user's identity.

Re: representatives, let's discuss in Slack. I'm not sure if we're on the same page here...

**Comment by igaponov** [comment]
`GET /api/v2/groups/{id}/responses` for group poll report

**Comment by petervojtek** [comment]
@igaponov , so I tried to run the GET request and this is the outcome:

```
Request URL:https://api-dev.powerli.ne/api/v2/groups/341/responses
Request Method:GET
Status Code:500 Internal Server Error

{"code":500,"message":"An exception occurred while executing 'SELECT u.firstName AS first_name, u.lastName AS last_name, u.address1, u.address2, u.city, u.state, u.country, u.zip, u.email, u.phone, u.bio, u.slogan, CASE WHEN u.facebook_id IS NOT NULL THEN 1 ELSE 0 END AS facebook, COUNT(f.id) AS followers, CASE WHEN a0.privacy = ? THEN \"Anonymous\" ELSE o0.value END AS `a`, CASE WHEN a1.privacy = ? THEN \"Anonymous\" ELSE o1.value END AS `discuss Section 1`, CASE WHEN a2.privacy = ? THEN \"Anonymous\" ELSE o2.value END AS `a`, CASE WHEN a3.privacy = ? THEN \"Anonymous\" ELSE o3.value END AS `ff`, CASE WHEN a4.privacy = ? THEN \"Anonymous\" ELSE o4.value END AS `.`, CASE WHEN a5.privacy = ? THEN \"Anonymous\" ELSE o5.value END AS `.`, CASE WHEN a6.privacy = ? THEN \"Anonymous\" ELSE o6.value END AS `.`, CASE WHEN a7.privacy = ? THEN \"Anonymous\" ELSE o7.value END AS `.`, CASE WHEN a8.privacy = ? THEN \"Anonymous\" ELSE o8.value END AS `.`, CASE WHEN a9.privacy = ? THEN \"Anonymous\" ELSE o9.value END AS `.`, CASE WHEN a10.privacy = ? THEN \"Anonymous\" ELSE o10.value END AS `.`, CASE WHEN a11.privacy = ? THEN \"Anonymous\" ELSE o11.value END AS `.`, CASE WHEN a12.privacy = ? THEN \"Anonymous\" ELSE o12.value END AS `.`, CASE WHEN a13.privacy = ? THEN \"Anonymous\" ELSE o13.value END AS `.`, CASE WHEN a14.privacy = ? THEN \"Anonymous\" ELSE o14.value END AS `.`, CASE WHEN a15.privacy = ? THEN \"Anonymous\" ELSE o15.value END AS `Poll 1`, CASE WHEN a16.privacy = ? THEN \"Anonymous\" ELSE o16.value END AS `4`, CASE WHEN a17.privacy = ? THEN \"Anonymous\" ELSE o17.value END AS `D2`, CASE WHEN a18.privacy = ? THEN \"Anonymous\" ELSE o18.value END AS `group discussion with video attachment`, CASE WHEN a19.privacy = ? THEN \"Anonymous\" ELSE o19.value END AS `image in attachment 1`, CASE WHEN a20.privacy = ? THEN \"Anonymous\" ELSE o20.value END AS `Show picked answer - poll1`, CASE WHEN a21.privacy = ? THEN \"Anonymous\" ELSE o21.value END AS `sf`, CASE WHEN a22.privacy = ? THEN \"Anonymous\" ELSE o22.value END AS `sdf`, CASE WHEN a23.privacy = ? THEN \"Anonymous\" ELSE o23.value END AS `sdf`, CASE WHEN a24.privacy = ? THEN \"Anonymous\" ELSE o24.value END AS `poll with attachment 1`, CASE WHEN a25.privacy = ? THEN \"Anonymous\" ELSE o25.value END AS `sdf`, CASE WHEN a26.privacy = ? THEN \"Anonymous\" ELSE o26.value END AS `with photo attachment`, CASE WHEN a27.privacy = ? THEN \"Anonymous\" ELSE o27.value END AS `sf`, CASE WHEN a28.privacy = ? THEN \"Anonymous\" ELSE o28.value END AS `Test 55`, CASE WHEN a29.privacy = ? THEN \"Anonymous\" ELSE o29.value END AS `desc`, CASE WHEN a30.privacy = ? THEN \"Anonymous\" ELSE o30.value END AS `lets discuss`, CASE WHEN a31.privacy = ? THEN \"Anonymous\" ELSE o31.value END AS `.`, CASE WHEN a32.privacy = ? THEN \"Anonymous\" ELSE o32.value END AS `attach video`, CASE WHEN a33.privacy = ? THEN \"Anonymous\" ELSE o33.value END AS `test www.yahoo.com image`, CASE WHEN a34.privacy = ? THEN \"Anonymous\" ELSE o34.value END AS `test 5`, CASE WHEN a35.privacy = ? THEN \"Anonymous\" ELSE o35.value END AS `test video on iphone`, CASE WHEN a36.privacy = ? THEN \"Anonymous\" ELSE o36.value END AS `video test 2`, CASE WHEN a37.privacy = ? THEN \"Anonymous\" ELSE o37.value END AS `111` FROM user u LEFT JOIN users_follow f ON f.user_id = u.id LEFT JOIN poll_answers a0 ON a0.user_id = u.id AND a0.question_id = ? LEFT JOIN poll_answers a1 ON a1.user_id = u.id AND a1.question_id = ? LEFT JOIN poll_answers a2 ON a2.user_id = u.id AND a2.question_id = ? LEFT JOIN poll_answers a3 ON a3.user_id = u.id AND a3.question_id = ? LEFT JOIN poll_answers a4 ON a4.user_id = u.id AND a4.question_id = ? LEFT JOIN poll_answers a5 ON a5.user_id = u.id AND a5.question_id = ? LEFT JOIN poll_answers a6 ON a6.user_id = u.id AND a6.question_id = ? LEFT JOIN poll_answers a7 ON a7.user_id = u.id AND a7.question_id = ? LEFT JOIN poll_answers a8 ON a8.user_id = u.id AND a8.question_id = ? LEFT JOIN poll_answers a9 ON a9.user_id = u.id AND a9.question_id = ? LEFT JOIN poll_answers a10 ON a10.user_id = u.id AND a10.question_id = ? LEFT JOIN poll_answers a11 ON a11.user_id = u.id AND a11.question_id = ? LEFT JOIN poll_answers a12 ON a12.user_id = u.id AND a12.question_id = ? LEFT JOIN poll_answers a13 ON a13.user_id = u.id AND a13.question_id = ? LEFT JOIN poll_answers a14 ON a14.user_id = u.id AND a14.question_id = ? LEFT JOIN poll_answers a15 ON a15.user_id = u.id AND a15.question_id = ? LEFT JOIN poll_answers a16 ON a16.user_id = u.id AND a16.question_id = ? LEFT JOIN poll_answers a17 ON a17.user_id = u.id AND a17.question_id = ? LEFT JOIN poll_answers a18 ON a18.user_id = u.id AND a18.question_id = ? LEFT JOIN poll_answers a19 ON a19.user_id = u.id AND a19.question_id = ? LEFT JOIN poll_answers a20 ON a20.user_id = u.id AND a20.question_id = ? LEFT JOIN poll_answers a21 ON a21.user_id = u.id AND a21.question_id = ? LEFT JOIN poll_answers a22 ON a22.user_id = u.id AND a22.question_id = ? LEFT JOIN poll_answers a23 ON a23.user_id = u.id AND a23.question_id = ? LEFT JOIN poll_answers a24 ON a24.user_id = u.id AND a24.question_id = ? LEFT JOIN poll_answers a25 ON a25.user_id = u.id AND a25.question_id = ? LEFT JOIN poll_answers a26 ON a26.user_id = u.id AND a26.question_id = ? LEFT JOIN poll_answers a27 ON a27.user_id = u.id AND a27.question_id = ? LEFT JOIN poll_answers a28 ON a28.user_id = u.id AND a28.question_id = ? LEFT JOIN poll_answers a29 ON a29.user_id = u.id AND a29.question_id = ? LEFT JOIN poll_answers a30 ON a30.user_id = u.id AND a30.question_id = ? LEFT JOIN poll_answers a31 ON a31.user_id = u.id AND a31.question_id = ? LEFT JOIN poll_answers a32 ON a32.user_id = u.id AND a32.question_id = ? LEFT JOIN poll_answers a33 ON a33.user_id = u.id AND a33.question_id = ? LEFT JOIN poll_answers a34 ON a34.user_id = u.id AND a34.question_id = ? LEFT JOIN poll_answers a35 ON a35.user_id = u.id AND a35.question_id = ? LEFT JOIN poll_answers a36 ON a36.user_id = u.id AND a36.question_id = ? LEFT JOIN poll_answers a37 ON a37.user_id = u.id AND a37.question_id = ? LEFT JOIN poll_options o0 ON a0.option_id = o0.id LEFT JOIN poll_options o1 ON a1.option_id = o1.id LEFT JOIN poll_options o2 ON a2.option_id = o2.id LEFT JOIN poll_options o3 ON a3.option_id = o3.id LEFT JOIN poll_options o4 ON a4.option_id = o4.id LEFT JOIN poll_options o5 ON a5.option_id = o5.id LEFT JOIN poll_options o6 ON a6.option_id = o6.id LEFT JOIN poll_options o7 ON a7.option_id = o7.id LEFT JOIN poll_options o8 ON a8.option_id = o8.id LEFT JOIN poll_options o9 ON a9.option_id = o9.id LEFT JOIN poll_options o10 ON a10.option_id = o10.id LEFT JOIN poll_options o11 ON a11.option_id = o11.id LEFT JOIN poll_options o12 ON a12.option_id = o12.id LEFT JOIN poll_options o13 ON a13.option_id = o13.id LEFT JOIN poll_options o14 ON a14.option_id = o14.id LEFT JOIN poll_options o15 ON a15.option_id = o15.id LEFT JOIN poll_options o16 ON a16.option_id = o16.id LEFT JOIN poll_options o17 ON a17.option_id = o17.id LEFT JOIN poll_options o18 ON a18.option_id = o18.id LEFT JOIN poll_options o19 ON a19.option_id = o19.id LEFT JOIN poll_options o20 ON a20.option_id = o20.id LEFT JOIN poll_options o21 ON a21.option_id = o21.id LEFT JOIN poll_options o22 ON a22.option_id = o22.id LEFT JOIN poll_options o23 ON a23.option_id = o23.id LEFT JOIN poll_options o24 ON a24.option_id = o24.id LEFT JOIN poll_options o25 ON a25.option_id = o25.id LEFT JOIN poll_options o26 ON a26.option_id = o26.id LEFT JOIN poll_options o27 ON a27.option_id = o27.id LEFT JOIN poll_options o28 ON a28.option_id = o28.id LEFT JOIN poll_options o29 ON a29.option_id = o29.id LEFT JOIN poll_options o30 ON a30.option_id = o30.id LEFT JOIN poll_options o31 ON a31.option_id = o31.id LEFT JOIN poll_options o32 ON a32.option_id = o32.id LEFT JOIN poll_options o33 ON a33.option_id = o33.id LEFT JOIN poll_options o34 ON a34.option_id = o34.id LEFT JOIN poll_options o35 ON a35.option_id = o35.id LEFT JOIN poll_options o36 ON a36.option_id = o36.id LEFT JOIN poll_options o37 ON a37.option_id = o37.id GROUP BY u.id' with params [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 426, 449, 450, 451, 455, 456, 457, 458, 459, 460, 461, 462, 466, 467, 468, 472, 473, 476, 485, 486, 487, 488, 489, 490, 491, 493, 495, 499, 500, 501, 502, 505, 509, 512, 514, 515, 516, 517]:\n\nSQLSTATE[HY000]: General error: 1116 Too many tables; MySQL can only use 61 tables in a join","errors":null}
```

**Comment by jterps08** [comment]
Let me know @devotebest if you need me to send a quick mockup for the item detail screen button via Slack please...

**Comment by devotebest** [comment]
@jterps08 
yeah, I need mockup, that would be very helpful to understand this issue.

**Comment by jterps08** [comment]
@igaponov can you provide all applicable endpoints for the reports that have been developed? I assume we have response reports for **single** polls, petitions, fundraisers, and events? as well as a membership roster report? I do not expect **group** polls/petitions/fundraisers/events reports...

**Comment by jterps08** [comment]
mockup for @devotebest ... **BUTTON A** this is just the Event-type Item Detail Screen, but ti would need to appear on poll-type, leader petition-type, fundraisers, and event-type item detail screens... Tapping the button will display alert message "Do you want to download the report for this item?" Yes / Cancel. 

**BUTTON B** Clicking the "Export Data" button (should be re-labeled to "Export Roster") displays alert "Do you want to download the detailed Membership Roster?"

![image](https://cloud.githubusercontent.com/assets/7582457/22234826/24e22374-e1c9-11e6-8000-4482ae004dbe.png)

**Comment by igaponov** [comment]
Endpoints:
Single poll: [`GET /api/v2/polls/{id}/responses`](https://api-dev.powerli.ne/api-doc#get--api-v2-polls-{id}-responses)
Membership roster: [`GET /api/v2/groups/{id}/members`](https://api-dev.powerli.ne/api-doc#get--api-v2-groups-{id}-members)

**Comment by jterps08** [comment]
@igaponov For "single poll", will that work for leader petitions, fundraisers, and events?

**Comment by igaponov** [comment]
@jterps08 yes, on backend leader petitions, fundraisers and events are the same entity, so have the same reports.

**Comment by devotebest** [comment]
@jterps08 
**but it would need to appear on poll-type, leader petition-type, fundraisers, and event-type item detail screens... Tapping the button will display alert message "Do you want to download the report for this item?" Yes / Cancel.

Clicking the "Export Data" button (should be re-labeled to "Export Roster") displays alert "Do you want to download the detailed Membership Roster?"**

I didn't get about this.

We need two buttons show different two alerts?

**Comment by devotebest** [comment]
@jterps08 
could you check my last comment?

**Comment by jterps08** [comment]
No... They are two different buttons located in two different places.

The first button (labeled Button A in comment above) appears on Item Detail Screen as indicated in mockup. Should be blue, not black. Appears for polls, leader-petitions, fundraisers, events... 

The second button is in the Manage Group screen at the bottom in Reports tab. It exists already. It's just not connected.

**Comment by devotebest** [comment]
After call API and get CSV format, and we need to download csv file to android device?

**Comment by jterps08** [comment]
Yes

**Comment by devotebest** [comment]
@jterps08 
we need to add "Button A" on Question Detail page also?

**Comment by jterps08** [comment]
This is Button A mockup. It's circled in red. This button A needs to appear on Item Detail screen for polls, leader-petitions, fundraisers, and events. 

![image](https://cloud.githubusercontent.com/assets/7582457/22336568/af1a6374-e3b0-11e6-9b4b-848e03e5d093.png)

**Comment by jterps08** [comment]
This is Button B

![screenshot_20170126-102153](https://cloud.githubusercontent.com/assets/7582457/22337245/81a87a14-e3b2-11e6-8e93-197ac3f43d55.png)

**Comment by devotebest** [comment]
Okay, got it.
and Export Data should be `Export Roster`

**Comment by jterps08** [comment]
Yes please - thank you!

**Comment by jterps08** [comment]
The file is not download to my device. I have an Android 6.01 Galaxy S7... I have not tested with older Android or iOS device. 

I would have expected a file to be downloaded to my device. 

![image](https://cloud.githubusercontent.com/assets/7582457/22343635/dc22e894-e3c6-11e6-8cc5-795b24197f60.png)

**Comment by devotebest** [comment]
Right now
Please check this folder
/Android/data/ne.powerli/files

**Comment by jterps08** [comment]
Okay - we're going to need a different experience for the user. I would expect to see the download appear in the system notification drawer as any normal file download would. Otherwise, the user does not have easy access to it.

**Comment by jterps08** [comment]
maybe the file-transfer plugin for cordova? I'm not sure exactly.

**Comment by jterps08** [comment]
Do you understand what I mean @devotebest? Something like below... 

![image](https://cloud.githubusercontent.com/assets/7582457/22345576/9f5243b2-e3ce-11e6-97c8-85d68522492c.png)

**Comment by igaponov** [comment]
Use
https://api-dev.powerli.ne/api-doc#get--api-v2-groups-{id}-members-link
https://api-dev.powerli.ne/api-doc#get--api-v2-polls-{id}-responses-link
to get a link for download

**Comment by devotebest** [comment]
@igaponov 
I am getting errors now
*https://api-dev.powerli.ne/api-public/files/8fd4a501-e7f7-4268-af9e-7654dce42c75*

**Comment by jterps08** [comment]
The file will only last for a couple minutes before it's deleted @devotebest - please confirm you're checking in time.

**Comment by jterps08** [comment]
@devotebest - The button works for group leaders. I need to test that it appears on polls, leader petitions, fundraisers, and events. I will only be able to test later this evening. Please confirm that the button 1) appears only for group leaders, and 2) appears on all four of those item types. 

@igaponov - The report that is being generated needs to reference the Group Permissions. Only the fields that the leader has permission to see (e.g. Name, Country, Responses) should appear in the report. Sorry if this wasn't clear. The leader should not see information s/he does not have permission to see. I believe columns A-J in the CSV generated are all controlled based on Group Permissions. Everything else (K through whatever) can be provided without permission as this information is available on the platform anyway.

**Comment by jterps08** [comment]
@devotebest - Can you confirm that this button appears on polls, leader petitions, fundraisers, and events? Please confirm that the button 1) appears only for group leaders, and 2) appears on all four of those item types.

**Comment by jterps08** [comment]
@igaponov Are you sure that all private fields are being hidden? Group G2017 only has permissions for Name, Country, Responses, yet I am sseing more user information in the downloaded fundraiser report. Can you please investigate?

@devotebest We still need to ensure that this button appears on polls, leader petitions, fundraisers, and events *only* for authorized group leaders.

**Comment by devotebest** [comment]
@jterps08 
what means `authorized group leaders` and `group leaders`?

**Comment by jterps08** [comment]
authorized group leaders are people who are either group owners or group managers for a group.

User A is a group manager (or owner) of Group 1, so he should see the Report button on the Item Detail Screen for any leader content (polls/petitions/fundraisers/events) created in Group 1.

User A is not a group manager (or owner) for Group 2, so he cannot see Report button on the Item Detail Screen for any leader content (polls/petitions/fundraisers/events) created in Group 2.

User B is a not a group manager (or owner) of Group 1, so he cannot see the Report button on the Item Detail Screen for any leader content (polls/petitions/fundraisers/events) created in Group 1.

 Only authorized group leaders can create leader content, so only authorized group leaders should be able to see reports.  Let me know if this makes sense.  It should be fairly simple to tell whether or not the user is an authorized group leader for any group's item...

**Comment by devotebest** [comment]
@igaponov 
how to check current user is authorized group leader for any group's item?

**Comment by jterps08** [comment]
@devotebest You can probably inspect the code for when the user tries to create a new fundraiser vs a new user post... it will only show certain groups to the user (only the groups the user is authorized to create leader content in)...

**Comment by devotebest** [comment]
there is function to check owner about Petition
userPetition.ownedByCurrentUser()
petition.ownedByCurrentUser()

but didn't find out any method about `question` (payment request and event).

trying to find other way now

**Comment by devotebest** [comment]
@igaponov 
please answer if you read this comment.

**Comment by jterps08** [comment]
Look for the "Boost" icon (looks like a megaphone). It only appears on a post if the logged-in user is an authorized group leader in which the post was created. @devotebest

**Comment by igaponov** [comment]
@devotebest https://api-dev.powerli.ne/api-doc#get--api-v2-groups-{id} attribute `user_role` for logged in user can be `owner`, `manager`, `member` or null.

**Comment by devotebest** [comment]
@igaponov 
so we need to call that api for all types?

**Comment by devotebest** [comment]
and where is user_role field in response?

**Comment by jterps08** [comment]
Re-opening to address included fields issue: https://github.com/PowerlineApp/powerline-mobile/issues/410#issuecomment-276982542

**Comment by igaponov** [comment]
@jterps08 fixed

**Comment by jterps08** [comment]
for report button issue, reference this part here https://github.com/PowerlineApp/powerline-mobile/issues/410#issuecomment-275981756 @rudjdev

**Comment by rudjdev** [comment]
what can I reference here related with #524?
